bookkeeper-server/src/main/java/org/apache/bookkeeper/client/UpdateLedgerOp.java

```diff
@@ -81,174 +87,80 @@ public UpdateLedgerOp(final BookKeeper bkc, final BookKeeperAdmin admin) {
-            this.stop = stop;
-            this.issuedLedgerCnt = issuedLedgerCnt;
-            this.updatedLedgerCnt = updatedLedgerCnt;
-            this.outstandings = outstandings;
-            this.syncObj = syncObj;
-            this.progressable = progressable;
+            outstanding.add(promise);
+            promise.whenComplete((metadata, ex) -> outstanding.remove(promise));
         }
 
-        @Override
-        public void onFailure(Throwable th) {
-            LOG.error("Error updating ledger {}", ledgerId, th);
-            stop.set(true);
-            finishUpdateLedger();
-        }
-
-        @Override
-        public void onSuccess(Void obj) {
-            updatedLedgerCnt.incrementAndGet();
-            // may print progress
-            progressable.progress(updatedLedgerCnt.get(), issuedLedgerCnt.get());
-            finishUpdateLedger();
-        }
-
-        private void finishUpdateLedger() {
-            outstandings.remove(ledgerId);
-            if (outstandings.isEmpty() && stop.get()) {
-                LOG.info("Total number of ledgers issued={} updated={}", issuedLedgerCnt.get(), updatedLedgerCnt.get());
-                syncObj.countDown();
-            }
+        try {
+            CompletableFuture.allOf(outstanding.stream().toArray(CompletableFuture[]::new)).get();
+            LOG.info("Total number of ledgers issued={} updated={}",
+                     issuedLedgerCnt.get(), updatedLedgerCnt.get());
+        } catch (ExecutionException e) {
+            LOG.info("Error in execution", e);
+            throw new IOException("Error executing update", e);
         }
     }
 
-    private static final class ReadLedgerMetadataCb implements GenericCallback<LedgerMetadata> {
-        final BookKeeper bkc;
-        final Long ledgerId;
-        final BookieSocketAddress curBookieAddr;
-        final BookieSocketAddress toBookieAddr;
-        SettableFuture<Void> future = SettableFuture.create();
-        public ReadLedgerMetadataCb(BookKeeper bkc, Long ledgerId, BookieSocketAddress curBookieAddr,
-                BookieSocketAddress toBookieAddr) {
-            this.bkc = bkc;
-            this.ledgerId = ledgerId;
-            this.curBookieAddr = curBookieAddr;
-            this.toBookieAddr = toBookieAddr;
+    private static LedgerMetadata replaceBookieInEnsembles(LedgerMetadata metadata,
+                                                           BookieSocketAddress oldBookieId,
+                                                           BookieSocketAddress newBookieId) {
+        LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+        for (Map.Entry<Long, ? extends List<BookieSocketAddress>> e : metadata.getEnsembles().entrySet()) {
+            List<BookieSocketAddress> newEnsemble = e.getValue().stream()
+                .map(b -> b.equals(oldBookieId) ? newBookieId : b)
+                .collect(Collectors.toList());
+            builder.withEnsembleEntry(e.getKey(), newEnsemble);
```

Comment:
I have a comment around `withEnsembleEntry` in last PR but since I didn't know how it was used so I didn't left the comment.

but here it is:

- since it is used for updating an existing ensemble, i prefer renaming `withEnsembleEntry` to `replaceEnsembleEntry` to make things explicit and clear
- if it is a replace operation, let's add a validation before put to make sure the ensemble exists. https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerMetadataBuilder.java#L94

Reply:
Will change to newEnsembleEntry and replaceEnsembleEntry with validation for each.